src/lib.rs

```diff
@@ -198,67 +287,200 @@ impl IRCompilerFactory {
-                self.me,
+                self.me.as_ptr(),
                 entry_point.as_ptr(),
-                input.me,
+                input.me.as_ptr(),
                 &mut error,
             )
-        };
+        });
 
         if error.is_null() {
             Ok(IRObject {
-                me: v,
+                me: v.unwrap(),
                 funcs: input.funcs.clone(),
             })
         } else {
-            panic!("{:?}", error);
+            let error = IRError::from_ptr(error, self.funcs.clone());
+            Err(CompilerError(error.code()))
+        }
+    }
+}
+
+struct IRError {
+    me: NonNull<bindings::IRError>,
+    funcs: Arc<bindings::metal_irconverter>,
+}
+
+impl Drop for IRError {
+    #[doc(alias = "IRErrorDestroy")]
+    fn drop(&mut self) {
+        unsafe { self.funcs.IRErrorDestroy(self.me.as_ptr()) };
+    }
+}
+
+impl IRError {
+    pub fn from_ptr(ptr: *mut bindings::IRError, funcs: Arc<bindings::metal_irconverter>) -> Self {
+        Self {
+            me: NonNull::new(ptr).unwrap(),
+            funcs,
+        }
+    }
+
+    #[doc(alias = "IRErrorGetCode")]
+    pub fn code(&self) -> IRErrorCode {
+        let code = unsafe { self.funcs.IRErrorGetCode(self.me.as_ptr()) };
+        match code {
+            0 => IRErrorCode::NoError,
+            1 => IRErrorCode::ShaderRequiresRootSignature,
+            2 => IRErrorCode::UnrecognizedRootSignatureDescriptor,
+            3 => IRErrorCode::UnrecognizedParameterTypeInRootSignature,
+            4 => IRErrorCode::ResourceNotReferencedByRootSignature,
+            5 => IRErrorCode::ShaderIncompatibleWithDualSourceBlending,
+            6 => IRErrorCode::UnsupportedWaveSize,
+            7 => IRErrorCode::UnsupportedInstruction,
+            8 => IRErrorCode::CompilationError,
+            9 => IRErrorCode::FailedToSynthesizeStageInFunction,
+            10 => IRErrorCode::FailedToSynthesizeStreamOutFunction,
+            11 => IRErrorCode::FailedToSynthesizeIndirectIntersectionFunction,
+            12 => IRErrorCode::UnableToVerifyModule,
+            13 => IRErrorCode::UnableToLinkModule,
+            14 => IRErrorCode::UnrecognizedDXILHeader,
+            15 => IRErrorCode::InvalidRaytracingAttribute,
+            16 => IRErrorCode::Unknown,
+            _ => panic!("Invalid error code"),
+        }
+    }
+
+    #[doc(alias = "IRErrorGetPayload")]
+    pub fn _payload(&self) -> CString {
+        unsafe {
+            // The documentation is inconsistent about this function.
+            // The docs say `You must cast this pointer to the appropriate error payload struct for the error code``
+            // but the example code just treats this as a char* like this: `printf("%s\n", (const char*)IRErrorGetPayload(pRootSigError));`
+            // Let's assume the example code is correct
+            let payload = self.funcs.IRErrorGetPayload(self.me.as_ptr()) as *mut i8;
+            CString::from_raw(payload)
```

Comment:
https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw

> Retakes ownership of a CString that was transferred to C via [CString::into_raw](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw).

So that's a no :)

Reply:
I'm not sure what to do with this one

Reply:
`CStr::from_ptr()`. See last commit.